openwisp-config/files/sbin/openwisp-update-config.lua

```diff
@@ -62,24 +64,46 @@ end
 -- update UCI configuration
 -- (overwrite or merge)
 
+stored = get_stored()
 -- remove config items added via openwisp-config
 if lfs.attributes(remote_config_dir, 'mode') == 'directory' then
     for file in lfs.dir(remote_config_dir) do
         local standard_path = standard_config_dir .. '/' .. file
         if lfs.attributes(standard_path, 'mode') == 'file' then
             for key, section in pairs(remote:get_all(file)) do
-                -- search section in the downloaded configuration
-                local section_check = check:get(file, section['.name'])
-                -- remove section from current configuration if not in downloaded configuration
-                if section_check == nil then
-                    utils.remove_uci_options(standard, file, section)
+                -- if section has been removed in the new configuration
+                if check:get(file, section['.name']) == nil then
+                    -- search section in the backup configuration
+                    local section_stored = stored:get_all(file, section['.name'])
+                    -- section is not in the backup configuration -> remove
+                    if section_stored == nil then
+                        utils.remove_uci_options(standard, file, section)
+                    -- section is in the backup configuration -> restore
+                    -- delete all options first
+                    else
+                        for option, value in pairs(section) do
+                            if not utils.starts_with_dot(option) then
+                                standard:delete(file, section['.name'], option)
```

Comment:
Hi @nemesisdesign @okraits
Here, if we are going to write the section, then what's the point of removing options before

Reply:
The exact section from the backup configuration needs to be restored, so all existing options are removed so the restore can start from a clean state.

Reply:
> The exact section from the backup configuration needs to be restored, so all existing options are removed so the restore can start from a clean state.

Thanks for clearing this :smile: 
I wrote some tests for this here, https://github.com/openwisp/openwisp-config/commit/a1e565b03e890d81b7db87f109bcf2da5f004dd4. I tried but wasn't to able update this pull request.